File: leetcode_problems/python/lc698_partition_to_k_equals_sum_substes.py
class Solution:
    def canPartitionKSubsets(self, nums: list[int], k: int) -> bool:
        def backtrack(i: int) -> bool:
            if i >= n:
                return all(s == t for s in subs)

            for isub in range(k):
                if subs[isub] + nums[i] <= t:
                    subs[isub] += nums[i]
                    if backtrack(i + 1):
                        return True
                    subs[isub] -= nums[i]

                if subs[isub] == 0:
                    break

            return False

        n = len(nums)
        nsum = sum(nums)
        if nsum <= 0 or nsum % k != 0:
            return False

        t = nsum // k
        if max(nums) > t:
            return False

        nums.sort(reverse=True)
        subs = [0] * k
        return backtrack(0)


if __name__ == "__main__":
    nums = [4, 3, 2, 3, 5, 2, 1]
    k = 4

    nums = [1, 2, 3, 4]
    k = 3
    print(Solution().canPartitionKSubsets(nums, k))

File: leetcode_problems/cpp/lc698_partition_to_k_equals_sum_substes.cpp
#include <vector>
#include <algorithm>
using std::vector;

class Solution
{
public:
    bool backtrack(int i, int t, vector<int> &nums, int subs[], int k)
    {
        if (i >= nums.size())
        {
            for (int isub = 0; isub < k; isub++)
            {
                if (subs[isub] != t)
                {
                    return false;
                }
            }
            return true;
        }

        for (int isub = 0; isub < k; isub++)
        {
            if (subs[isub] + nums[i] <= t)
            {
                subs[isub] += nums[i];
                if (backtrack(i + 1, t, nums, subs, k))
                {
                    return true;
                }
                subs[isub] -= nums[i];
            }

            if (subs[isub] == 0)
            {
                break;
            }
        }
        return false;
    }

    bool canPartitionKSubsets(vector<int> &nums, int k)
    {
        int n = nums.size();
        long long nsum = 0;
        for (int num : nums)
        {
            nsum += num;
        }
        if (nsum <= 0 || nsum % k != 0)
        {
            return false;
        }

        int t = nsum / k;
        if (*std::max_element(nums.begin(), nums.end()) > t)
        {
            return false;
        }

        std::sort(nums.rbegin(), nums.rend());
        int *subs = new int[n]();
        return backtrack(0, t, nums, subs, k);
    }
};